### PR TITLE
Archive lib refactoring

### PR DESCRIPTION
**Pull request by Holt59**
I started this to improve performance. This improves the performance quite a bit (switching from Qt IO to Win32 API in particular), but the huge boost was outside of MO2, it was Windows Defender... Since I had started refactoring, I just had to finish.

**Refactoring:**
- Switch usage of Qt IO stuff to Win32 API or filesystem. Most of the code is taken from the actual 7z implementation (`FileBase`, `FileOut`, `FileIn` in particular).
- Switch usage of `QString` for either `std::wstring` or `std::filesystem::path` with proper long path handling.
- Use `std::function` instead of the home-made callbacks.
- Use a dedicated `LogCallback` for logging instead of `qDebug()`. This callback can be set in MO2 to use `log::debug` (see the modorganizer PR).
- The progress callback now monitor extraction instead of... whatever it was monitoring before. See WIP.
- Expose the (reference) `vector` of `FileData` instead of a weird pointer + size. It’s the exact same interface, but with proper C++.
- Added some instrumentation stuff but these could be removed easily.

**Notes:**
- While the changes seems important, most of the stuff is actually taken from 7z, or simply replacing Qt stuff -> standard stuff, and this only impacts the installation manager on the MO2 side.
- I wanted this to be kind of independent of major stuff, so it only requires `fmt` right now (and 7z of course), which is why there some utility functions (e.g. for strings).

I tried to test everything as much as I could, but here is the shortlist of the « edge cases » I tested:
- Path to the extracted archive is longer than the default maximum length.
- Path to the folder where the archive is extracted is long than the default maximum length.
- Archive with or without passwords, with or without encrypted file names (see the short discussion on Discord).

### CODE REVIEW

**Comment by Al12rs** [comment]
Remember to update the readme for any important usage change. From what I can see this PR aims to make archive more independent from Mo2 so we should also make sure the instructions are clear for anyone else to use it.

File: README.md [hunk 1]
```diff
@@ -2,77 +2,214 @@
 
 # modorganizer-archive
 
-This module provides a wrapper round the 7zip 7z.dll allowing easy(ish) access to the contents of an archive.
+This module provides a wrapper round the 7zip `7z.dll` allowing easy(ish) access to the contents of an archive.
+
+As this is a DLL you need to load it as an archive.
+
+```cpp
+std::unique_ptr<Archive> CreateArchive();
```
Comment (Al12rs) [inline README.md:10]: move this under the class header

File: README.md [hunk 2]
```diff
@@ -2,77 +2,238 @@
...
+ *
+ * @return true if the archive was extracted, false otherwise.
+ */
+bool extract(std::wstring const &outputDirectory, ArchiveCallbacks::ProgressCallback progressCallback,
+             ArchiveCallbacks::FileChangeCallback fileChangeCallback, ArchiveCallbacks::ErrorCallback errorCallback)
+```
+
+All callbacks are optional, you can pass an empty `std::function` instead (either `nullptr` or `{}`). The purpose of the callbacks:
+
+- `progressCallback(float)` is called during extraction to notify progress. _Note to self: update._
```
Comment (Al12rs) [inline README.md:118]: There seem to be some Note to slef here left over

File: src/archive.h [hunk 3]
```diff
@@ -39,19 +36,86 @@ class QString;
...
+   */
+  virtual void addOutputFileName(std::wstring const& filepath) = 0;
+  
+  /**
+   * @brief Retrieve the list of filepaths to extract this entry to and clear the
+   *   internal list.
+   *
+   * @return the list of paths this entry should be extracted to.
+   */
+  virtual std::vector<std::wstring> getAndClearOutputFileNames() = 0;
```
Comment (Al12rs) [inline src/archive.h:64]: This seems very impractical for anyone else wanting to use this library. 
Adding a just getOutputFileNames() to just retrieve what is already there would probably beconvenient.
If we want to go the extra mile we could allow to pass a vector of filepahts directly to addOutPutFileName().

Comment (Holt59) [reply src/archive.h:64]: I kind of agree, I just kept the original functionality. I don't really mind changing it.

I'm not sure having `addOutputFileName` overloads is really useful - It's a two line call with a `for ()` loop, and I'm not sure you would often add multiple output names:

```cpp
for (auto &x: v) file->addOutputFileName(x);
```

And to be "clean", that would probably be a template on `It`, so the call would not be that short: 

```cpp
file->addOutputFileNames(std::begin(v), std::end(v));
```

BTW, if we're changing this, I'm probably going to rename `FileName` to `Path` or `FilePath`.

Comment (Holt59) [reply src/archive.h:64]: I can also directly return the `vector<std::wstring>&`, and people can do whatever they want. I'm not really against this honestly...

Comment (Al12rs) [reply src/archive.h:64]: Agreed for the vector overload not being that useful. 
For the fileName I was about to go check the implementation to see what it actually is, the comment says full path but my guess was just the path relative to the archive root?

Comment (Holt59) [reply src/archive.h:64]: Well, `getFileName` has nothing to do with `addOutputFileName` or `getAndClearOutputFileNames` actually since `getFileName` is the path in the archive. But yes, all path are relative: relative to the root of the archive or to the extraction folder.

Well, actually I don't really now how `getFileName` behave when the archive contains absolute path but that's the same behavior as before.

I was pretty sure to have documented this somewhere but I'll change the function documentations.

Comment (Holt59) [reply src/archive.h:64]: It's specified in the README actually, but better put it here also.

Comment (Al12rs) [reply src/archive.h:64]: > I can also directly return the `vector<std::wstring>&`, and people can do whatever they want. I'm not really against this honestly since there is no reason not to allow users to modify the vector of filepaths.

Nah I prefer encapsulation the operations on it in functions in case we ever need to make this threadsafe or change the behavior in some other way.

File: src/archive.h [hunk 4]
```diff
@@ -39,19 +36,86 @@ class QString;
...
+  enum class LogLevel {
+    Debug,
+    Info,
+    Warning,
+    Error
+  };
+
+  enum class ProgressType {
+    
+    // Indicates the progression in the archive. This may reach 100% way before the
```
Comment (Al12rs) [inline src/archive.h:90]: What does this mean? The progress in reading the archive during the extraction? What should this be used for?

Comment (Holt59) [reply src/archive.h:90]: Are you asking about the enumeration of about the `ARCHIVE` value? The enumeration is to select different progress type in `ProgressCallback`.

The `ARCHIVE` value is related to "reading" the archive during extraction, but I don't really now if it's `READING` (it's not clear at all in 7z "documentation"). It was what was used by MO2 prior to this PR to update progress, which is why for some archives, you either reached 99% very quickly and spent most of the time on the last %, or you reached 100% before the extraction was complete and then the progress dialog started disappearing / re-appearing in a loop until the extraction was complete.

I've let this because this can probably be useful sometimes. For instance, I may use this progress instead of `EXTRACTION` prior to installation because it might be more accurate in this case.

You'd have to test the PRs to see the difference.

Comment (Al12rs) [reply src/archive.h:90]: Yes I was asking what the archive flag meant vs extraction. From the comment it's not really clear if there is any use case for it, so I would add to that comment what you have found through testing, like the fact that it might be better for extracting just a few files.

Comment (Holt59) [reply src/archive.h]: I've updated the comment for both enumeration values.

File: src/archive.h (hunk 3 above)
Comment (Holt59) [reply src/archive.h]: I've updated `FileData` to use `path` instead of `name`, be clear in the comments that those are relative path (to the output folder), split `getAndClear` in `get` and `clear`.

Comment (Al12rs) [reply src/archive.h]: Nice work!
